Use Minecraft's system for recipes

ramidzkh opened the pull request:
Using Minecraft's system for recipes ensures easy syncing and for players to easily edit recipes through datapacks

---
- [X] Migrate from Ore Dictionary in favor of tags
- [X] Use Minecraft's JSON recipe system

ramidzkh committed:
Move gated recipes

ramidzkh committed:
ItemStack -> Ingredient (1/x)

ramidzkh committed:
ItemStack -> Ingredient (2/x)

ramidzkh committed:
Update CentrifugeRecipe

ramidzkh committed:
Use Minecraft's system for recipes

Nedelosk commented on src/main/java/forestry/factory/recipes/CarpenterRecipe.java:
```diff
@@ -55,4 +64,28 @@ public FluidStack getFluidResource() {
+		return id;
+	}
+
+	// TODO: Remove ShapedRecipeCustom. We cannot serialize this without ShapedRecipeCustom having a serializer
```
Is ShapedRecipeCustom still needed at all anymore?

ramidzkh replied:
I'm pretty sure it's not, so I'll start removing it

ramidzkh committed:
Revert "Move gated recipes"

I thought these were all written by hand

ramidzkh committed:
Remove ShapedRecipeCustom

ramidzkh committed:
Serialize CarpenterRecipe

ramidzkh committed:
Begin work on data generator recipe builders

ramidzkh committed:
Add license headers

ramidzkh committed:
Soldering recipes

ramidzkh committed:
Remove some methods

ramidzkh committed:
Make getRecipes use datapacks

ramidzkh committed:
Goodbye static collections!

ramidzkh committed:
Goodbye static in recipe managers!

ramidzkh committed:
Fix carpenters (kinda)

Nedelosk commented on src/main/java/forestry/api/recipes/ICarpenterManager.java:
```diff
@@ -54,4 +57,16 @@
+	 * @param CraftingInventory Present inventory
+	 * @return An optional carpenter recipe if any matches
+	 */
+	Optional<ICarpenterRecipe> findMatchingRecipe(RecipeManager recipeManager, FluidStack liquid, ItemStack item, IInventory CraftingInventory);
```
```suggestion
	Optional<ICarpenterRecipe> findMatchingRecipe(RecipeManager recipeManager, FluidStack liquid, ItemStack item, IInventory craftingInventory);
```

Nedelosk commented on src/main/java/forestry/factory/recipes/RecipeSerializers.java:
```diff
@@ -0,0 +1,88 @@
+		return object;
+	}
+
+	public static JsonObject item(ItemStack stack) {
```
Is there no vanilla / forge method for this?

ramidzkh replied:
Apparently not. I first thought of using Dynamics to convert the NBT item to a JSON object but I don't think ShapedRecipe.deserializeItem would work with it

Nedelosk commented on src/main/java/forestry/factory/recipes/RecipeSerializers.java:
```diff
@@ -0,0 +1,88 @@
+		return stack;
+	}
+
+	public static JsonObject serializeFluid(FluidStack fluid) {
```
Is there no vanilla / forge method for this?

ramidzkh replied:
I couldn't find anything; only for serializing to NBT writing to packet buffers

Nedelosk commented on src/main/java/forestry/api/recipes/IForestryRecipe.java:
```diff
@@ -5,5 +5,86 @@
+	@Deprecated
+	@Override
+	default String getGroup() {
+		return "Forestry";
```
Are all recipes groups uppercase?

ramidzkh replied:
Just had a proper look; they are "" if there is no group, or the group is picked from the recipe JSON

Nedelosk replied:
yea, but all the group names from the jsons are lowercase

ramidzkh committed:
Use camel casing for CraftingInventory

ramidzkh committed:
Quick reformatting

ramidzkh committed:
Remove ore dictionary from recipes?

ramidzkh committed:
Lowercase recipe group

ramidzkh committed:
AbstractCraftingProvider

ramidzkh committed:
Stuff

ramidzkh committed:
Merge remote-tracking branch 'upstream/dev-1.16' into recipes

ramidzkh committed:
Lazy FilteredTank predicates

ramidzkh committed:
Stuff

ramidzkh committed:
Nullable RecipeMananger

ramidzkh committed:
Make it compile!

TODO - Fix SqueezerContainerRecipe, SqueezerContainerRecipeManager
TODO - fix RecipeUtil#addFermenterRecipes

ramidzkh committed:
Fix SqueezerContainerRecipeManager

ramidzkh committed:
Fix SqueezerContainerRecipe

Its more of a Frankenstein now

Nedelosk commented:
Looks very nice!

ramidzkh committed:
Make adjust work

ramidzkh committed:
Don't use null everywhere

ramidzkh committed:
Don't make the containers spam the logs

ramidzkh committed:
Make the still work

ramidzkh committed:
Make the fermenter work

ramidzkh committed:
Remove the last mentions of the ore dictionary in recipes

ramidzkh committed:
Re-add some fermenter recipes

ramidzkh committed:
Implement fabricator recipe (de)serialization

ramidzkh committed:
Update copyright

ramidzkh commented:
All that's left to fix are carpenters, fabricators and worktables. No recipes have actually been converted to JSON (through data generation) yet too

ramidzkh committed:
Fix fluid capabilities with the bottler

ramidzkh committed:
Carpenters maybe

ramidzkh committed:
Delete RecipePair

ramidzkh committed:
More oredict removing

ramidzkh committed:
More oredict removing

ramidzkh committed:
Use Ingredient for squeezer recipes

Nedelosk committed:
Add Ingredient inventory support and fix worktable

ramidzkh committed:
Allow nullable fluids

ramidzkh committed:
Almost fix fabricators

ramidzkh committed:
Ingredients for fabricators

ramidzkh committed:
Fix fabricators

ramidzkh commented:
All JSON defined recipes should work now. Some code defined still won't work (fabricators and carpenters). This means all that's left is converting the code recipes to use recipe builders (which are not yet implemented). Other than that, I think this is ready for review

ramidzkh committed:
Modify existing builders

ramidzkh committed:
FabricatorRecipeBuilder and FermenterRecipeBuilder

ramidzkh committed:
All recipe builders

ramidzkh committed:
Remove advancements from recipe builders

Nedelosk commented on src/main/java/forestry/factory/tiles/TileCarpenter.java:
```diff
@@ -202,7 +200,7 @@ private boolean removeItemResources(boolean doRemove) {
 		NonNullList<ItemStack> craftingSets = InventoryUtil.getStacks(craftingInventory, InventoryGhostCrafting.SLOT_CRAFTING_1, InventoryGhostCrafting.SLOT_CRAFTING_COUNT);
 		IInventory inventory = new InventoryMapper(getInternalInventory(), InventoryCarpenter.SLOT_INVENTORY_1, InventoryCarpenter.SLOT_INVENTORY_COUNT);
-		return InventoryUtil.removeSets(inventory, 1, craftingSets, oreDicts, null, true, false, doRemove);
+		return InventoryUtil.removeSets(inventory, 1, craftingSets, null, true, false, doRemove);
```
Whats about using the ingridient version her too ?

Nedelosk commented on src/main/java/forestry/factory/tiles/TileFabricator.java:
```diff
@@ -228,9 +227,9 @@ private void craftResult() {
+	private boolean removeFromInventory(NonNullList<ItemStack> set, boolean doRemove) {
 		IInventory inventory = new InventoryMapper(this, InventoryFabricator.SLOT_INVENTORY_1, InventoryFabricator.SLOT_INVENTORY_COUNT);
-		return InventoryUtil.removeSets(inventory, 1, set, recipePair.getOreDictEntries(), null, true, false, doRemove);
+		return InventoryUtil.removeSets(inventory, 1, set, null, true, false, doRemove);
```
Same like the carpenter

Nedelosk commented on src/main/java/forestry/factory/tiles/TileSqueezer.java:
```diff
@@ -161,18 +161,40 @@ public boolean workCycle() {
-			if (currentRecipe != null && ItemStackUtil.containsSets(currentRecipe.getResources(), resources, true, false) > 0) {
+			boolean containsSets = false;
+
+			if (currentRecipe != null) {
```
Should be replaced with  simulated "consumeIngredients" I thing

ramidzkh committed:
Fix more ingredient stuff

ramidzkh commented:
I think everything is done except the recipes themselves 🎉

ramidzkh committed:
Port centrifuge recipes

ramidzkh committed:
Port fabricator smelting recipes

ramidzkh committed:
Port fermenter recipes

ramidzkh committed:
Port moistener recipes

ramidzkh committed:
Port squeezer container recipes

ramidzkh committed:
Port still recipe

ramidzkh committed:
Port squeezer recipes

ramidzkh committed:
Port hygroregulator recipe

ramidzkh committed:
Datagen recipes

ramidzkh committed:
Use some tags

Fixes FabricatorSmeltingRecipeBuilder
Makes the build task actually work

ramidzkh committed:
Use more tags

ramidzkh committed:
Move

ramidzkh committed:
Port most of the carpenter recipes + datagen

ramidzkh committed:
Port most of the fabricator recipes

ramidzkh committed:
Consistency

ramidzkh committed:
Datagen fabricator recipes

ramidzkh committed:
More recipes

ramidzkh committed:
Carpenter recipe output overrides

ramidzkh committed:
Everything is now data generated

ramidzkh committed:
Minor fixes

ramidzkh committed:
Major fixes

ramidzkh committed:
Complete it all

ramidzkh commented:
Everything seems to be complete now.
---
I need people to playtest it and try to find any bugs relating to the recipes themselves (machines not accepting a certain recipe, the recipes not completing correctly, etc)

Minecraft 1.16.1
Forge `1.16.1-32.0.63`
Download: [Extract this ZIP for the mod jar](https://github.com/ForestryMC/ForestryMC/files/6040775/forestry_extract_me.zip)

ramidzkh committed:
Allow fabricators to smelt

ramidzkh committed:
Add fruit squeezer recipes
